File: crates/detrix-lsp/tests/fixtures/go/pure_chain.go
// Test fixture: Pure function call chain (3 levels deep)
//
// Call chain: CalculateTotal -> ComputeSubtotal -> ComputeItemPrice
// All functions are pure (no side effects).
package main

import (
	"fmt"
	"math"
)

// Item represents a single item with quantity and price
type Item struct {
	Quantity  int
	UnitPrice float64
}

// ComputeItemPrice is Level 3: Pure function - computes item price using math operations.
func ComputeItemPrice(quantity int, unitPrice float64) float64 {
	return math.Round(float64(quantity)*unitPrice*100) / 100
}

// ComputeSubtotal is Level 2: Pure function - calls ComputeItemPrice for each item.
func ComputeSubtotal(items []Item) float64 {
	total := 0.0
	for _, item := range items {
		total += ComputeItemPrice(item.Quantity, item.UnitPrice)
	}
	return total
}

// CalculateTotal is Level 1: Pure function - calls ComputeSubtotal and applies tax.
func CalculateTotal(items []Item, taxRate float64) float64 {
	subtotal := ComputeSubtotal(items)
	tax := math.Floor(subtotal*taxRate*100) / 100
	return subtotal + tax
}

// FormatCurrency is a pure function - formats amount as currency string.
func FormatCurrency(amount float64) string {
	return fmt.Sprintf("$%.2f", amount)
}

// ValidateItems is a pure function - validates items slice.
func ValidateItems(items []Item) bool {
	if len(items) == 0 {
		return false
	}
	for _, item := range items {
		if item.Quantity < 0 || item.UnitPrice < 0 {
			return false
		}
	}
	return true
}

File: crates/detrix-lsp/tests/fixtures/pure_chain.py
"""
Test fixture: Pure function call chain (3 levels deep)

Call chain: calculate_total -> compute_subtotal -> compute_item_price
All functions are pure (no side effects).
"""
import math


def compute_item_price(quantity: int, unit_price: float) -> float:
    """Level 3: Pure function - computes item price using math operations."""
    return round(quantity * unit_price, 2)


def compute_subtotal(items: list[tuple[int, float]]) -> float:
    """Level 2: Pure function - calls compute_item_price for each item."""
    total = 0.0
    for quantity, price in items:
        total += compute_item_price(quantity, price)
    return total


def calculate_total(items: list[tuple[int, float]], tax_rate: float) -> float:
    """Level 1: Pure function - calls compute_subtotal and applies tax."""
    subtotal = compute_subtotal(items)
    tax = math.floor(subtotal * tax_rate * 100) / 100
    return subtotal + tax


# Additional pure helper functions for testing
def format_currency(amount: float) -> str:
    """Pure function - formats amount as currency string."""
    return f"${amount:.2f}"


def validate_items(items: list) -> bool:
    """Pure function - validates items list."""
    return len(items) > 0 and all(
        isinstance(item, tuple) and len(item) == 2 for item in items
    )
